### PR TITLE
Add MapGroup

### LINKED ISSUE #36007: [Epic]: Improve minimal api routing with grouping
## Background and Motivation
The new minimal approach in asp.net core 6 is good. I mean it's great. But I don't this will be useful in a real web application. Because this doesn't have an easy way to integrate route grouping. With route grouping, minimal actions can be separated into multiple files easily with the help of normal or static(since c# doesn't support modules like f#) classes without using MVC or Razor pages.

## Proposed API
```diff
namespace Microsoft.AspNetCore.Builder;

+ public class RouteGroup : IApplicationBuilder, IEndpointRouteBuilder
+ {
+     // Needed members to implement those interfaces
+ }

public class MinimalActionEndpointRouteBuilderExtensions
{
+     public static void IncludeRouteGroup(this IEndpointRouteBuilder endpoints, string pattern, RouteGroup routeGroup);
}
```

## Usage Examples
```C#
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

RouteGroup usersRoutes = new();
usersRoutes.MapGet("{id}", (int id) => GetUserById(id));

app.IncludeRouteGroup("/api/users/", usersRoutes);
```
You can separate these route actions to multiple files easily. RouteGroup would become more useful if in the future c# gets support for top-level functions(not just for Program.cs).

### PR DESCRIPTION
**Pull request by halter73**
This PR add a new MapGroup() extension method to `IEndpointRouteBuilder`. This returns a new builder for defining groups of endpoints with a common prefix that implements both the `IEndpointRouteBuilder` and `IEndpointConventionBuilder` interfaces. This can be used to add endpoints with the given group prefix, and to customize those endpoints using conventions.

#### Sample Usage

```csharp
var app = WebApplication.Create(args);

var group = app.MapGroup("/todos");
group.MapGet("/", (int id, TodoDb db) => db.ToListAsync());
group.MapGet("/{id}", (int id, TodoDb db) => db.GetAsync(id));
group.MapPost("/", (Todo todo, TodoDb db) => db.AddAsync(todo));

var nestedGroup = group.MapGroup("/{org}");
nestedGroup.MapGet("/", (string org, TodoDb db) => db.Filter(todo => todo.Org == org).ToListAsync());
nestedGroup.RequireCors("AllowAll");

// This applies to all endpoints defined on group and nestedGroup
group.RequireAuthorization();

// ...
```

Fixes #36007

### CODE REVIEW

File: src/Http/Routing/src/Patterns/RoutePatternFactory.cs [hunk 1]
```diff
@@ -1084,6 +1084,108 @@ public static RoutePatternParameterPolicyReference ParameterPolicy(string parame
         return ParameterPolicyCore(parameterPolicy);
     }
 
+    internal static RoutePattern Combine(RoutePattern left, RoutePattern right)
```
Comment (davidfowl) [inline src/Http/Routing/src/Patterns/RoutePatternFactory.cs:1087]: This looks allocatey

Comment (halter73) [reply src/Http/Routing/src/Patterns/RoutePatternFactory.cs:1087]: I'm surprised you say that considering the optimizations. The `RawText` string and the `RoutePattern` itself are the only allocations in most cases.

File: src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs [hunk 2]
```diff
@@ -28,6 +28,41 @@ public static class EndpointRouteBuilderExtensions
...
+    /// <summary>
+    /// Create a <see cref="GroupRouteBuilder"/> for defining endpoints all prefixed with the specified <paramref name="prefixPattern"/>.
+    /// </summary>
+    /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/> to add the group to.</param>
+    /// <param name="prefixPattern">The pattern that prefixes all routes in this group.</param>
+    /// <returns>
+    /// A <see cref="GroupRouteBuilder"/> that is both a <see cref="IEndpointRouteBuilder"/> and a <see cref="IEndpointConventionBuilder"/>.
+    /// The same builder can be used to add endpoints with the given <paramref name="prefixPattern"/>, and to customize those endpoints using conventions.
+    /// </returns>
+    public static GroupRouteBuilder MapGroup(this IEndpointRouteBuilder endpoints, RoutePattern prefixPattern)
```
Comment (davidfowl) [inline src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:52]: Do we have other public APIs that accept a RoutePattern?

**Review by davidfowl — APPROVED** [review]
I'd like to see one test with MVC/SignalR/Health checks, something that's not the ModelDataSource.

Oh ALSO, I'd like a test that fires the change token for a data source that fires change tokens and it working.

File: src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs [hunk 3]
```diff
@@ -28,6 +28,41 @@ public static class EndpointRouteBuilderExtensions
...
+    /// <param name="prefixPattern">The pattern that prefixes all routes in this group.</param>
+    /// <returns>
+    /// A <see cref="GroupRouteBuilder"/> that is both a <see cref="IEndpointRouteBuilder"/> and a <see cref="IEndpointConventionBuilder"/>.
+    /// The same builder can be used to add endpoints with the given <paramref name="prefixPattern"/>, and to customize those endpoints using conventions.
+    /// </returns>
+    public static GroupRouteBuilder MapGroup(this IEndpointRouteBuilder endpoints, string prefixPattern) =>
+        endpoints.MapGroup(RoutePatternFactory.Parse(prefixPattern ?? throw new ArgumentNullException(nameof(prefixPattern))));
+
+    /// <summary>
+    /// Create a <see cref="GroupRouteBuilder"/> for defining endpoints all prefixed with the specified <paramref name="prefixPattern"/>.
```
Comment (Kahbazi) [inline src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:44]: nit:
```suggestion
    /// Creates a <see cref="GroupRouteBuilder"/> for defining endpoints all prefixed with the specified <paramref name="prefixPattern"/>.
```

File: src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs [hunk 4]
```diff
@@ -28,6 +28,41 @@ public static class EndpointRouteBuilderExtensions
     private static readonly string[] DeleteVerb = new[] { HttpMethods.Delete };
     private static readonly string[] PatchVerb = new[] { HttpMethods.Patch };
 
+    /// <summary>
+    /// Create a <see cref="GroupRouteBuilder"/> for defining endpoints all prefixed with the specified <paramref name="prefixPattern"/>.
```
Comment (Kahbazi) [inline src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:32]: nit:
```suggestion
    /// Creates a <see cref="GroupRouteBuilder"/> for defining endpoints all prefixed with the specified <paramref name="prefixPattern"/>.
```

File: src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs [hunk 5]
```diff
@@ -28,6 +28,41 @@ public static class EndpointRouteBuilderExtensions
...
+    /// </summary>
+    /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/> to add the group to.</param>
+    /// <param name="prefixPattern">The pattern that prefixes all routes in this group.</param>
+    /// <returns>
+    /// A <see cref="GroupRouteBuilder"/> that is both a <see cref="IEndpointRouteBuilder"/> and a <see cref="IEndpointConventionBuilder"/>.
```
Comment (BrennanConroy) [inline src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:37]:
```suggestion
    /// A <see cref="GroupRouteBuilder"/> that is both an <see cref="IEndpointRouteBuilder"/> and an <see cref="IEndpointConventionBuilder"/>.
```

File: src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs [hunk 6]
```diff
@@ -28,6 +28,41 @@ public static class EndpointRouteBuilderExtensions
...
+    public static GroupRouteBuilder MapGroup(this IEndpointRouteBuilder endpoints, string prefixPattern) =>
+        endpoints.MapGroup(RoutePatternFactory.Parse(prefixPattern ?? throw new ArgumentNullException(nameof(prefixPattern))));
+
+    /// <summary>
+    /// Create a <see cref="GroupRouteBuilder"/> for defining endpoints all prefixed with the specified <paramref name="prefixPattern"/>.
+    /// </summary>
+    /// <param name="endpoints">The <see cref="IEndpointRouteBuilder"/> to add the group to.</param>
+    /// <param name="prefixPattern">The pattern that prefixes all routes in this group.</param>
+    /// <returns>
+    /// A <see cref="GroupRouteBuilder"/> that is both a <see cref="IEndpointRouteBuilder"/> and a <see cref="IEndpointConventionBuilder"/>.
```
Comment (BrennanConroy) [inline src/Http/Routing/src/Builder/EndpointRouteBuilderExtensions.cs:49]:
```suggestion
    /// A <see cref="GroupRouteBuilder"/> that is both an <see cref="IEndpointRouteBuilder"/> and an <see cref="IEndpointConventionBuilder"/>.
```

File: src/Http/Routing/src/GroupRouteBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            get
+            {
+                var list = new List<Endpoint>();
+
+                foreach (var dataSource in _groupRouteBuilder._dataSources)
+                {
+                    foreach (var endpoint in dataSource.Endpoints)
+                    {
+                        // Endpoint does not provide a RoutePattern but RouteEndpoint does. So it's impossible to apply a prefix custom Endpoints.
```
Comment (BrennanConroy) [inline src/Http/Routing/src/GroupRouteBuilder.cs:74]:
```suggestion
                        // Endpoint does not provide a RoutePattern but RouteEndpoint does. So it's impossible to apply a prefix for custom Endpoints.
```

File: src/Http/Routing/src/Patterns/RoutePatternFactory.cs [hunk 8]
```diff
@@ -1084,6 +1084,108 @@ public static RoutePatternParameterPolicyReference ParameterPolicy(string parame
...
+            }
+            return combinedList;
+        }
+
+        // Technically, the ParameterPolicies could probably be merged because it's a list, but it makes little sense to add policy
+        // for the same parameter in both the left and right part of the combined pattern. Defaults and Required values cannot be
+        // merged because the `TValue` is `object?`, but over-setting a Default or RequiredValue (which may not be in the parameter list)
+        // seems okay as long as the values are the same for a given key in both the left and right pattern. There's already similar logic
+        // in PatternCore for when defaults come from both the `defaults` and `segements` param. `requiredValues` cannot be defined in
+        // `segements` so there's no equivalent to merging these until now.
```
Comment (BrennanConroy) [inline src/Http/Routing/src/Patterns/RoutePatternFactory.cs:1126]:
```suggestion
        // in PatternCore for when defaults come from both the `defaults` and `segments` param. `requiredValues` cannot be defined in
        // `segments` so there's no equivalent to merging these until now.
```

**Review by BrennanConroy — APPROVED** [review]
Just some comment cleanup nits

**Comment by davidfowl** [comment]
Do you have a test where we call MapGroup multiple times with the same prefix? I know we're not keeping track of the prefix but it shouldn't matter unless you have duplicate routes right?

**Pull request merged**